### PR TITLE
Fix s3 download and touch

### PR DESCRIPTION
**Pull request by icewind1991**
- Fix constructing the headers for s3 stream download
- use non empty files when `touch`ing an objectstore file.

### CODE REVIEW

**Comment by schiessle** [comment]
It seems to work, although I still get some error messages in the nextcloud log:

during the initial setup I get a lot of

````
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:52:47+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A5): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:53:55+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A11): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:53:58+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A13): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:54:00+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A15): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:54:02+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A18): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:54:04+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A21): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"fhpm3fMo5ZC1U8UPiXQw","level":3,"time":"2017-11-09T15:54:06+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"POST","url":"\/master\/index.php","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A23): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"PcD03Sw9N7f8sDHfRPWN","level":3,"time":"2017-11-09T15:54:10+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/apps\/files\/","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A29): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
...
````

after the setup finished and I upload I file I get this error messages;

````
{"reqId":"gnoi73Fl1XmuSYvW6d75","level":3,"time":"2017-11-09T15:55:55+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2Ffragen.limux.txt&c=5a047a89a75c2&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A126): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"gnoi73Fl1XmuSYvW6d75","level":3,"time":"2017-11-09T15:55:57+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2Ffragen.limux.txt&c=5a047a89a75c2&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A127): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#68","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
````

maybe that's not a problem, we just try to access a file/preview which is not there yet. Still it is bad if we spam the nextcloud.log for every file operation.

**Comment by icewind1991** [comment]
I only get the errors when I first try the install on master, and then switch to this PR.

So I think this is caused by an inconsistent state from the first failed install.

**Comment by schiessle** [comment]
I can reproduce it. I switch first to your branch, delete the database (mariadb) and create a new/empty one. Delete the s3 bucket and then do the initial setup.

Before I perform the initial setup I already add to the config.php following, so that the object storage is configures right from the beginning:

````
<?php
$CONFIG = array (
  'installed' => false,
  'cache_path' => '/home/schiesbn/repos/nextcloud/server/tmp',
  'objectstore' => 
  array (
    'class' => 'OC\\Files\\ObjectStore\\S3',
    'arguments' => 
    array (
      'bucket' => 'nextcloud-schiesbn2',
      'autocreate' => false,
      'key' => 'my-s3-key',
      'secret' => 'my-s3-secret',
      'use_path_style' => true,
    ),
  ),
);
````
I also have all the example files in the skeleton directory: https://github.com/nextcloud/example-files. Maybe this makes a difference, if more files are created during initial setup/login?

If I navigate after the installation into the "Photos" directory I get this errors:

````
{"reqId":"fuOeeVEwIBumN151fchE","level":3,"time":"2017-11-11T10:52:19+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/apps\/theming\/img\/core\/filetypes\/image.svg?v=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A124): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"CodhvKSWoUSP1zpfQNH0","level":3,"time":"2017-11-11T10:52:21+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2FCoast.jpg&c=5a06d5563c802&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A127): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"UcreQYabeqMcxYgmyiCX","level":3,"time":"2017-11-11T10:52:23+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2FHummingbird.jpg&c=5a06d55c949a8&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A128): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"apku8Y7S5XA4mhULVM6p","level":3,"time":"2017-11-11T10:52:23+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2FNut.jpg&c=5a06d5594e3d1&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A129): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"CodhvKSWoUSP1zpfQNH0","level":3,"time":"2017-11-11T10:52:25+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2FCoast.jpg&c=5a06d5563c802&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A130): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"UcreQYabeqMcxYgmyiCX","level":3,"time":"2017-11-11T10:52:28+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2FHummingbird.jpg&c=5a06d55c949a8&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A131): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"apku8Y7S5XA4mhULVM6p","level":3,"time":"2017-11-11T10:52:28+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2FNut.jpg&c=5a06d5594e3d1&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A132): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
````

If I upload a non-empty text file, I get this errors:

````
{"reqId":"xXHG7DtmbYQp0RJvMKzz","level":3,"time":"2017-11-11T10:59:11+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2Ffragen.limux.txt&c=5a06d7fbec9ba&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A135): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
{"reqId":"xXHG7DtmbYQp0RJvMKzz","level":3,"time":"2017-11-11T10:59:13+00:00","remoteAddr":"::1","user":"admin","app":"PHP","method":"GET","url":"\/master\/index.php\/core\/preview.png?file=%2FPhotos%2Ffragen.limux.txt&c=5a06d7fbec9ba&x=32&y=32&forceIcon=0","message":"fopen(https:\/\/s3.eu-west-1.amazonaws.com\/nextcloud-schiesbn2\/urn%3Aoid%3A136): failed to open stream: HTTP request failed! HTTP\/1.1 404 Not Found\r\n at \/home\/schiesbn\/repos\/nextcloud\/server\/master\/lib\/private\/Files\/ObjectStore\/S3ObjectTrait.php#70","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"13.0.0.5"}
````

At the end all files are accessible and the preview are generated and shown. But it seems like in the process there is a situation where we try to access a non existing file.

**Comment by schiessle** [comment]
If I just set "installed" to false and cleanup the database and reinstall Nextcloud I don't get any error messages. So maybe that's the difference between our tests? That I delete the bucket first so that a new bucket needs to be created and all files needs to be uploaded again and you install it with a already existing bucket which contains already files from previous tries?

**Review by MorrisJobke — APPROVED** [review]
Tested and works here 👍

**Comment by schiessle** [comment]
@MorrisJobke you can't reproduce the 404 on preview generation? Did you tried it the way I described it? Real Amazon S3 backend? bucket doesn't exist before installation?

**Comment by MorrisJobke** [comment]
> @MorrisJobke you can't reproduce the 404 on preview generation? Did you tried it the way I described it? Real Amazon S3 backend? bucket doesn't exist before installation?

As of now I tried it with a fakes3 bucket, but an empty one. AFAIK there is no support for a prefilled bucket. I just then uploaded some documents, photos and movies. All were played back/shown correctly.

**Comment by schiessle** [comment]
The strange thing is also. If I debug it the fopen which fails comes from here: https://github.com/nextcloud/server/blob/master/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L301 As you can see we do a file_exists() first. Still the fopen fails and returns false so that we continue with `$source=false`

Maybe a timing issue, that the file is already added to the file cache but not uploaded, that would explain why file_exists() return true but fopen fails

**Comment by schiessle** [comment]
@icewind1991 so maybe the timing issue i described. Did you tried it already the way I described it?

**Comment by icewind1991** [comment]
@schiessle I can't reproduce the issue, imo we should merge this and look into the issue seperatly

**Comment by schiessle** [comment]
@icewind1991 ok, let's merge it and backport it for now

**Review by schiessle — APPROVED** [review]

**Comment by schiessle** [comment]
@icewind1991 can you create a backport? Thanks!

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 1]
```diff
@@ -364,7 +364,7 @@ public function touch($path, $mtime = null) {
 			$fileId = $this->getCache()->put($path, $stat);
 			try {
 				//read an empty file from memory
```
Comment (nickvergessen) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php]: not correct anymore

**Pull request merged**